app/master/cluster_master.py

```diff
@@ -109,46 +109,38 @@ def connect_slave(self, slave_url, num_executors):
         """
         Connect a slave to this master.
 
-        If this slave has already been registered with the master and isn't working on a build, make sure that slave
-        is marked as alive and ready to accept builds.
-
         :type slave_url: str
         :type num_executors: int
         :return: The response with the slave id of the slave.
         :rtype: dict[str, str]
         """
+        # If a slave had previously been connected, and is now being reconnected, the cleanest way to resolve this
+        # bookkeeping is for the master to forget about the previous slave instance and start with a fresh instance.
         if slave_url in self._all_slaves_by_url:
-            self._logger.warning('Slave on {} requested to connect to master, even though already connected', slave_url)
-            slave = self._all_slaves_by_url.get(slave_url)
-            slave.set_is_alive(True)
-
-            # If the slave has made a request to reconnect to the master, it's safe to say that whatever
-            # subjob/build that slave was working on at the time is no longer running. In order to prevent
-            # that build from hanging indefinitely (as the slave will never respond to the master with the
-            # subjob results for that build) we must cancel the build.
-            if slave.current_build_id is not None:
-                self._logger.warning('Newly reconnected slave {} was working on build {}, cancelling build.',
-                                     slave_url, slave.current_build_id)
+            self._logger.warning('Slave on {} requested to connect to master, even though previously connected. ' +
+                                 'Removing existing slave instance from the master\'s bookkeeping.', slave_url)
+            removed_slave = self._all_slaves_by_url.pop(slave_url)
```

Comment:
We might want to leave the old slave in the dict and just overwrite it with the new one. I'm wondering if there might be a tiny race between this line and when we add the new one below -- if some other thread tries to access the slave in `_all_slaves_by_url` it will get an exception. (I don't have an exact case for this in mind, it's just a feeling.)